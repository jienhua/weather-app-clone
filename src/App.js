import React, { Component } from 'react';
import './App.css';

import Plot from './Plot.js';

import { connect } from 'react-redux';

import { 
    changeLocation,
    setSelectedTemp,
    setSelectedDate,
    fetchData
} from './actions';

export class App extends Component {

    constructor () {
        super();
        this.onPlotClick = this.onPlotClick.bind(this);
    };

    fetchData (evt) {
        evt.preventDefault();
        let location = encodeURIComponent(this.props.location);

        let urlPrefix = 'http://api.openweathermap.org/data/2.5/forecast?q=';
        let urlSuffix = '&APPID=c45463aa2e7b9dfb9d79c3cb9b5b6c14&units=metric';
        let url = urlPrefix + location + urlSuffix;

        this.props.dispatch(fetchData(url));
    };
    
    changeLocation (evt) {
        this.props.dispatch(changeLocation(evt.target.value));
    };

    onPlotClick (data){
        if (data.points) {
            // let number = data.points[0].pointNumber;
            this.props.dispatch(setSelectedDate(data.points[0].x));
            this.props.dispatch(setSelectedTemp(data.points[0].y));
        }
    };

    render() {
        let currentTemp = 'not loaded yet';
        let { selected, data, location, dates, temps } = this.props;
        if (data.get('list')) {
            currentTemp = data.getIn(['list', '0', 'main', 'temp']);
        }
        return (
            <div>
                <h1>Weather</h1>
                <form onSubmit={(event)=>this.fetchData(event)}>
                    <label>I want to know the weather for
                        <input placeholder={'City, Country'} 
                               type='text' 
                               value={location}
                               onChange={(event)=>this.changeLocation(event)}/>
                    </label>
                </form>
     
                {(data.get('list')) ? (
                <div className='wrapper'>
                    <p className='temp-wrapper'>
                        <span className='temp'>{ selected.get('temp') ? selected.get('temp') : currentTemp }</span>
                        <span className='temp-symbol'>Â°C</span>
                        <span className="temp-date">
                            { selected.get('temp') ? selected.get('date') : ''}
                        </span>
                    </p>
                    <h2>Forecast</h2>
                    <Plot
                        xData={dates}
                        yData={temps}
                        onPlotClick={this.onPlotClick}
                        type='scatter'
                    />
                </div>
                ) : null}
            </div>
        );
    }
}


function mapStateToProps(state) {
    return {
        location: state.get('location'),
        selected: state.get('selected'),
        data: state.get('data'),
        dates: state.get('dates'),
        temps: state.get('temps')
    };
    // return state.toJS();
}


// export default App;
export default connect(
    mapStateToProps
)(App);