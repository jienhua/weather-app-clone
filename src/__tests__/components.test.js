import React from 'react';
import renderer from 'react-test-renderer';
import { App } from '../App';
import { fromJS } from 'immutable';
import Plot from '../Plot.js';

describe('components', function () {
	describe('<App />', function () {
		it('renders correctly', function () {
			let tree = renderer.create(
				<App location='' 
					 selected={fromJS({})}
					 data={fromJS({})}
					 dates={fromJS([])}
					 temps={fromJS([])} />).toJSON();
			expect(tree).toMatchSnapshot();
		});
	});

	describe('<Plot />', function () {
		global.Plotly = {
			newPlot: () => {}
		};
		global.document = {
	    	getElementById: function() { return {
	        	on: function() {}
	      	}}
	    };
		it('renders correctly', function () {
			const tree = renderer.create(<Plot xData={fromJS({})} yData={fromJS({})}/>).toJSON();
			expect(tree).toMatchSnapshot();
		});
	});
});