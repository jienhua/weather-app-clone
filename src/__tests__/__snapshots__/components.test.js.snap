// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components <App /> renders correctly 1`] = `
<div>
  <h1>
    Weather
  </h1>
  <form
    onSubmit={[Function]}
  >
    <label>
      I want to know the weather for
      <input
        onChange={[Function]}
        placeholder="City, Country"
        type="text"
        value=""
      />
    </label>
  </form>
</div>
`;
